### PR TITLE
Add plugin for Final Fantasy XIV

### PR DESCRIPTION
**Pull request by EmperorArthur**
This is the first draft of a positional audio plugin for Final Fantasy XIV.  It probably requires more work before being merge ready.  I used the l4d2 plugin as a base, with a few patterns found in the wow plugin.

I'd appreciate some feedback, especially since I'm pretty sure I've messed up the coordinate system.  I couldn't find any documentation on how to debug this thing aside from littering the code with `cout` statements.

The plugin does compile on Linux, though it's not set to in this pull request.  I'm in the process of setting up a Windows build environment, so I haven't had a chance to do any testing.

What do you all like about this code?  What do you dislike, and what should I change?

### CODE REVIEW

**Comment by mkrautz** [comment]
Hi @EmperorArthur,
Do you run FFXIV on Wine?

I haven't yet reviewed the PR, but here's my WIP branch for wine plugin support, if you're interested:
https://github.com/mkrautz/mumble/commits/wine-plugins-v5

File: plugins/plugins.pro [hunk 1]
```diff
@@ -12,7 +12,7 @@ SUBDIRS = link
 DIST = plugins.pri
 
 win32 {
-	SUBDIRS += aoc arma2 bf1942 bf2 bf3 bf2142 bfbc2 bfheroes bf4_x86 blacklight borderlands borderlands2 breach cod2 cod4 cod5 codmw2 codmw2so cs css dods dys etqw tf2 gmod gtaiv gw hl2dm insurgency jc2 l4d l4d2 lol lotro ql rl sr sto ut2004 ut3 ut99 wolfet wow
+	SUBDIRS += aoc arma2 bf1942 bf2 bf3 bf2142 bfbc2 bfheroes bf4_x86 blacklight borderlands borderlands2 breach cod2 cod4 cod5 codmw2 codmw2so cs css dods dys etqw tf2 gmod gtaiv gw hl2dm insurgency jc2 l4d l4d2 lol lotro ql rl sr sto ut2004 ut3 ut99 wolfet wow ffxiv
```
Comment (davidebeatrici) [inline plugins/plugins.pro]: _ffxiv_ should be after _etqw_, to keep the alphabetical order.

File: plugins/ffxiv/ffxiv.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	camera_pos[0] = camera_pos_corrector[0];
+	camera_pos[1] = camera_pos_corrector[2];
+	camera_pos[2] = camera_pos_corrector[1];
+
+	avatar_front[0] = std::cos(avatar_heading) ;
+	avatar_front[1] = std::sin(avatar_heading);
+	avatar_front[2] = 0;
+
+	if(camera_is_free) {
+		camera_front[0] = -std::cos(camera_heading) ;
```
Comment (davidebeatrici) [inline plugins/ffxiv/ffxiv.cpp]: Space after the bracket.

File: plugins/ffxiv/ffxiv.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+const procptr32_t map_id_address       = 0x10210B0;
+const procptr32_t camera_ptr           = 0x1045C40;
+const procptr32_t avatar_ptr           = 0x10468EC;
+const procptr32_t state_address        = 0x1048C60;
+//const procptr32_t servername_address = 0;
+// Char struct offsets
+const procptr32_t identity_offset       = 48;  // Name
+const procptr32_t avatar_pos_offset     = 160; // X, Z, Y
+const procptr32_t avatar_heading_offset = 176; // Heading (-pi to pi)
+//Camera struct offsets
```
Comment (davidebeatrici) [inline plugins/ffxiv/ffxiv.cpp]: This comment should be "// Camera struct offsets".

File: plugins/ffxiv/ffxiv.cpp [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+// Y is North/South, South is increasing values
+// X is East/West, West is increasing values
+// Heading is in radians, 0 faces south
+
+// Memory addresses
+const procptr32_t map_id_address       = 0x10210B0;
+const procptr32_t camera_ptr           = 0x1045C40;
+const procptr32_t avatar_ptr           = 0x10468EC;
+const procptr32_t state_address        = 0x1048C60;
+//const procptr32_t servername_address = 0;
```
Comment (davidebeatrici) [inline plugins/ffxiv/ffxiv.cpp]: If the server name isn't available I would just remove it.

File: plugins/ffxiv/ffxiv.cpp [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		}
+
+		return true; // This tells Mumble to ignore all vectors.
+	}
+
+	// Begin context
+	std::ostringstream ocontext;
+	ocontext << "{";
+
+	// Server name
```
Comment (davidebeatrici) [inline plugins/ffxiv/ffxiv.cpp]: If the server name isn't available I would just remove it.

**Comment by davidebeatrici** [comment]
Hi @EmperorArthur,

You're not obliged to setup a full environment to compile plugins, see the [Wiki page](http://wiki.mumble.info/wiki/Pluginguide#Out_of_tree_build_.28doesn.27t_need_a_Mumble_build_environment.29) for a small guide on how to do it.

To test plugins we have a dedicated program called [MumblePAHelper](https://github.com/mumble-voip/mumble-pahelper), you can download it from [here](https://dl.mumble.info/MumblePAHelper-winx64-2016-07-16.zip) or build it yourself if you want the latest version; we still have to setup a builder for it. :wink:

File: plugins/ffxiv/ffxiv.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	for (int i=0;i<3;i++) {
+		avatar_pos[i] = avatar_front[i] = avatar_top[i] = camera_pos[i] = camera_front[i] = camera_top[i] = 0.0f;
+	}
+
+	// Boolean value to check if game addresses retrieval is successful
+	bool ok;
+	// Create containers to stuff our raw data into, so we can convert it to Mumble's coordinate system
+	float avatar_pos_corrector[3], camera_pos_corrector[3], avatar_heading, camera_heading;
+	// Values for extra features
+//  char servername[50];
```
Comment (mkrautz) [inline plugins/ffxiv/ffxiv.cpp]: Delete?

File: plugins/ffxiv/ffxiv.pro [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2005-2016 The Mumble Developers. All rights reserved.
+# Use of this source code is governed by a BSD-style license
+# that can be found in the LICENSE file at the root of the
+# Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+include(../plugins.pri)
+
+TARGET		= ffxiv
+SOURCES		= ffxiv.cpp
```
Comment (mkrautz) [inline plugins/ffxiv/ffxiv.pro]: It's fine.

File: plugins/ffxiv/ffxiv.cpp [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+#include <cmath>
+
+// Offset values can be obtained from:
+//   http://xivapp.com/api/structures?patchVersion=latest&platform=x86
+// Y is North/South, South is increasing values
+// X is East/West, West is increasing values
+// Heading is in radians, 0 faces south
+
+// Memory offsets
```
Comment (EmperorArthur) [inline plugins/ffxiv/ffxiv.cpp]: Ok, i was worried about it after @davidebeatrici mentioned them being mixed.  It's the same in the l4d2 plugin.

Will work on fixing.

**Comment by EmperorArthur** [comment]
Good news, this compiles successfully on MSVC 2010, 2013, and 2015.  I know because I installed all three of them yesterday.

Bad news (for me),  MumblePAHelper and Mumble snapshot don't recognize any compiled plugin.  At this point I'm stuck.

**Comment by davidebeatrici** [comment]
You must use the same compiler (and architecture) for Mumble/MumblePAHelper and the plugins.

**Comment by mkrautz** [comment]
For Mumble 1.3.x, that's MSVC 2013, for Mumble 1.2.x, it's MSVC2010.

**Comment by EmperorArthur** [comment]
Ok, the problem was everything, including MumblePAHelper, is compiled for X64, but the wiki didn't mention that.  That should probably be fixed.

One final question before this thing is ready.  Does this sound correct?

> With the character facing South, `Avatar Front` is `X: 1, Y: 0, Z: 0`
> With the character facing East, `Avatar Front` is `X: 0, Y: 0, Z: 1`
> With the character facing South `Avatar_Pos_X` increases while going West. 
> With the character facing South `Avatar_Pos_Z` increases while going South.
> With the character facing South `Avatar_Pos_Y` increases while going Up.

South was chosen as the reference direction since heading (in radians) is 0 pointing South.

**Comment by mkrautz** [comment]
Seems wrong that the coordinate system for avatar_front is different from avatar_pos.. (South is X is one, in the other?)

Also, how is avatar_front's Z ever anything but 0 when it's set to 0 here? https://github.com/mumble-voip/mumble/pull/2653/files#diff-be1da379944da5c11f9a33c6658fcc3aR128

**Comment by EmperorArthur** [comment]
I was confused about the coordinate system.  Looking at the code for the manual placement plugin helped straighten things out.

`actor_front` Y is always zero.  The player can not tilt up and down.  There may be motions that look like it, but those animations do not affect the player's coordinates.

This code works, but I believe I can move things to North aligned instead of South aligned.  Also, would you like me to squash all commits and rebase to master?

side note:  I really wish the manual placement plugin was an actual plugin.  It would have made experimenting, and understanding the coordinate system so much easier.

File: plugins/ffxiv/ffxiv.cpp [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+		oidentity << "Player: null";
+	}
+
+	oidentity << "}";
+	identity = oidentity.str();
+	// End identity
+
+	/*
+	Mumble | Game (in L hand coordinate system)
+	X      | -X
```
Comment (EmperorArthur) [inline plugins/ffxiv/ffxiv.cpp]: I may be able to swap this to X, -Y, Z and save all the negatives in the azimuth calculations.

File: plugins/ffxiv/ffxiv.cpp [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+	// Avatar pointer
+	procptr32_t avatar_address = peekProc<procptr32_t>(pModule + avatar_ptr);
+	if (!avatar_address) return false;
+
+	// Camera pointer
+	procptr32_t camera_address = peekProc<procptr32_t>(pModule + camera_ptr);
+	if (!camera_address) return false;
+
+	// Peekproc and assign game addresses to our containers, so we can retrieve positional data
+	ok = peekProc(pModule + state_offset, &state, 1) && // Magical state value: 0 or 255 when in main menu and 1 when in-game.
```
Comment (EmperorArthur) [inline plugins/ffxiv/ffxiv.cpp]: Tabs or Spaces in aligning things here?

File: plugins/ffxiv/ffxiv.cpp (hunk 9 above)
Comment (mkrautz) [inline plugins/ffxiv/ffxiv.cpp]: I have no opinion. As long as the coordinate system is mapped sensibly onto the Mumble coordinate system, I'm happy.

**Comment by mkrautz** [comment]
> I was confused about the coordinate system. Looking at the code for the manual placement plugin helped straighten things out.
>
>actor_front Y is always zero. The player can not tilt up and down. There may be motions that look like it, but those animations do not affect the player's coordinates.
>
>This code works, but I believe I can move things to North aligned instead of South aligned.

Sounds like a good idea to me. I'm not sure what other plugins do, though!
Not a top priority, as long as the coordinate system is consistent.

>Also, would you like me to squash all commits and rebase to master?

Yes, please!

>side note: I really wish the manual placement plugin was an actual plugin. It would have made experimenting, and understanding the coordinate system so much easier.

It was, until very recently. We made it part of Mumble because of the Qt UI... Perhaps we can do something similar in MumblePAHelper, but use the link plugin?

File: plugins/ffxiv/ffxiv.cpp (hunk 10 above)
Comment (EmperorArthur) [inline plugins/ffxiv/ffxiv.cpp]: Ok, I'm aligning all the `&& //` so I'll use tabs for the indentation, and spaces to align those.

**Comment by EmperorArthur** [comment]
Here's, what should be, the final version.  Ready to merge, hopefully.

There is a 64 bit version of the game, but I will submit that as a separate pull request.

**Comment by davidebeatrici** [comment]
Looking good to me. :wink:

Does the DirectX 11 version of the game use a different executable?

**Comment by mkrautz** [comment]
Landed. Thanks a bunch! :-)

**Comment by EmperorArthur** [comment]
Thanks for all your help.  I enjoy Mumble, and appreciate everything you all do.

**Comment by davidebeatrici** [comment]
You're welcome! 
Thank you very much for your contribution! :smile:

**Pull request merged**